WalletWasabi/Crypto/Mac.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Text;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto
+{
+	public class MAC : IEquatable<MAC>
+	{
+		private MAC(Scalar t, GroupElement U, GroupElement V)
+		{
+			this.t = CryptoGuard.NotZero(nameof(t), t);
+			this.U = CryptoGuard.NotNullOrInfinity(nameof(U), U);
+			this.V = CryptoGuard.NotNullOrInfinity(nameof(V), V);
+		}
+
+		public Scalar t { get; }
+		public GroupElement U { get; }
+		public GroupElement V { get; }
+
+		public static bool operator == (MAC a, MAC b) => a.Equals(b);
+		public static bool operator != (MAC a, MAC b) => !a.Equals(b);
+
+		public bool Equals(MAC? other) =>
+			(this, other) switch
+			{
+				(null, null) => true,
+				(null, {}) => false,
+				({}, null) => false,
+				({} a , {} b) when Object.ReferenceEquals(a, b) => true, 
+				({} a , {} b) => (a.t, a.U, a.V) == (b.t, b.U, b.V)
+			};
+
+		public override bool Equals(object? obj) =>
+			Equals(obj as MAC);
+
+		public override int GetHashCode() =>
```

Comment:
what is this used for?

Reply:
This is something you have to implement when your type overrides `Equals` what is what you have to implement when your class is `IEquatable`. Basically it is to compute a hash that c# types will use when you store your instances in a `HashSet` or `Dictionary`. It is plumbing...